code/game/dna/dna2.dm

```diff
@@ -55,35 +54,74 @@ GLOBAL_LIST_EMPTY(bad_blocks)
 	new_dna.blood_type = blood_type
 	new_dna.real_name = real_name
 	new_dna.species = new species.type
-	for(var/b=1;b<=DNA_SE_LENGTH;b++)
-		new_dna.SE[b]=SE[b]
-		if(b<=DNA_UI_LENGTH)
-			new_dna.UI[b]=UI[b]
-	new_dna.UpdateUI()
-	new_dna.UpdateSE()
+	for(var/b in 1 to DNA_SE_LENGTH)
+		new_dna.SE[b] = SE[b]
+	for(var/b in 1 to DNA_UI_LENGTH)
+		new_dna.UI[b] = UI[b]
+	for(var/b in 1 to DNA_RP_LENGTH)
+		new_dna.RP[b] = RP[b]
+	new_dna.UpdateDNA(DNA_ALL)
 	return new_dna
 
 ///////////////////////////////////////
-// UNIQUE IDENTITY
+// UPDATE and RESET types of DNA
 ///////////////////////////////////////
 
-// Create random UI.
-/datum/dna/proc/ResetUI(defer = FALSE)
-	for(var/i=1,i<=DNA_UI_LENGTH,i++)
-		switch(i)
-			if(DNA_UI_SKIN_TONE)
-				SetUIValueRange(DNA_UI_SKIN_TONE, rand(1, 220), 220, 1) // Otherwise, it gets fucked
-			else
-				UI[i]=rand(0,4095)
-	if(!defer)
-		UpdateUI()
-
-/datum/dna/proc/ResetUIFrom(mob/living/carbon/human/character)
-	// INITIALIZE!
-	ResetUI(1)
-	// Hair
-	// FIXME:  Species-specific defaults pls
-	var/obj/item/organ/external/head/H = character.get_organ("head")
+/datum/dna/proc/ResetDNA(dna_type, defer = FALSE)
+	ASSERT(dna_type >= 0)
+	ASSERT(dna_type <= 3)
+	switch(dna_type)
+		if(DNA_UI)	// Create a random UI
+			for(var/i in 1 to DNA_UI_LENGTH)
+				if(i == DNA_UI_SKIN_TONE)
+					SetDNAValueRange(DNA_UI_SKIN_TONE, rand(1, 220), 220, DNA_UI, TRUE)	// Otherwise, it gets screwed up
+				else
+					UI[i] = rand(0, 4095)
+			if(!defer)
+				UpdateDNA(DNA_UI)
+		if(DNA_SE)	// "Zeroes out" all of the SE Blocks
+			for(var/i in 1 to (DNA_SE_LENGTH - 1))
+				SetDNAValue(i, rand(1, 1024), TRUE, DNA_SE)
+			UpdateDNA(DNA_SE)
+		if(DNA_RP)
+			for(var/i in 1 to DNA_RP_LENGTH)
+				SetDNAValue(i, rand(1, 1024), TRUE, DNA_RP)
```

Comment:
Ordering here looks like it should be:
`SetDNAValue(i, rand(1, 1024), DNA_RP, TRUE)`

Reply:
Fascinating.  I am impressed by myself. How did I miss these things?
I, myself, was looking at this the other day and was thinking, "Huh. That order looks wrong" and, yet, I couldn't see it. Thank you!